libraries/AP_OpticalFlow/AP_OpticalFlow_CXOF.cpp

```diff
@@ -0,0 +1,201 @@
+    return sensor;
+}
+
+// initialise the sensor
+void AP_OpticalFlow_CXOF::init()
+{
+    // sanity check uart
+    if (uart == nullptr) {
+        return;
+    }
+
+    // open serial port with baud rate of 19200
+    uart->begin(19200);
+
+    last_frame_us = AP_HAL::micros();
+}
+
+// read latest values from sensor and fill in x,y and totals.
+void AP_OpticalFlow_CXOF::update(void)
+{
+    // sanity check uart
+    if (uart == nullptr) {
+        return;
+    }
+
+    // record gyro values as long as they are being used
+    if (gyro_sum_count < 1000) {
+        const Vector3f& gyro = AP::ahrs_navekf().get_gyro();
+        gyro_sum.x += gyro.x;
+        gyro_sum.y += gyro.y;
+        gyro_sum_count++;
+    }
+
+    // sensor values
+    int32_t x_sum = 0;
+    int32_t y_sum = 0;
+    uint16_t qual_sum = 0;
+    uint16_t count = 0;
+
+    // read any available characters in the serial buffer
+    int16_t nbytes = uart->available();
+    while (nbytes-- > 0) {
+        int16_t r = uart->read();
+        if (r < 0) {
+            continue;
+        }
+        uint8_t c = (uint8_t)r;
+        // if buffer is empty and this byte is header, add to buffer
+        if (buf_len == 0) {
+            if (c == CXOF_HEADER) {
+                buf[buf_len++] = c;
+            }
+        } else {
+            // add character to buffer
+            buf[buf_len++] = c;
+
+            // if buffer has 9 items try to decode it
+            if (buf_len >= CXOF_FRAME_LENGTH) {
+                // check last character matches footer
+                if (buf[buf_len-1] != CXOF_FOOTER) {
+                    buf_len = 0;
+                    continue;
+                }
+
+                // decode package
+                int16_t x_raw = (int16_t)((uint16_t)buf[3] << 8) | buf[2];
+                int16_t y_raw = (int16_t)((uint16_t)buf[5] << 8) | buf[4];
+
+                // add to sum of all readings from sensor this iteration
+                count++;
+                x_sum += x_raw;
+                y_sum += y_raw;
+                qual_sum += buf[7];
+
+                // clear buffer
+                buf_len = 0;
+            }
+        }
+    }
+
+    // return without updating state if no readings
+    if (count == 0) {
+        return;
+    }
+
+    struct OpticalFlow::OpticalFlow_state state {};
+    state.device_id = 0x43; // 'C'
+
+    // average surface quality scaled to be between 0 and 255
+    state.surface_quality = (constrain_int16(qual_sum / count, 64, 78) - 64) * 255 / 14;
+
+    // calculate dt
+    uint64_t this_frame_us = uart->receive_time_constraint_us(CXOF_FRAME_LENGTH);
+    if (this_frame_us == 0) {
+        // for HAL that cannot estimate arrival time in serial buffer use current time
+        this_frame_us = AP_HAL::micros();
+    }
+    float dt = (this_frame_us - last_frame_us) * 1.0e-6;
+    last_frame_us = this_frame_us;
+
+    if ((dt > 0.0f) && (dt < CXOF_TIMEOUT_SEC)) {
```

Comment:
is_positive ?

Reply:
txs for the review! fixed in a follow up commit.